crates/utils/src/rate_limit/mod.rs

```diff
@@ -221,14 +252,45 @@ where
   }
 }
 
-fn get_ip(conn_info: &ConnectionInfo) -> IpAddr {
-  IpAddr(
-    conn_info
-      .realip_remote_addr()
-      .unwrap_or("127.0.0.1:12345")
-      .split(':')
-      .next()
-      .unwrap_or("127.0.0.1")
-      .to_string(),
-  )
+fn get_ip(conn_info: &ConnectionInfo) -> Ipv6Addr {
+  conn_info
+    .realip_remote_addr()
+    .and_then(parse_ip)
+    .unwrap_or(Ipv4Addr::new(127, 0, 0, 1).to_ipv6_mapped())
+}
+
+fn parse_ip(addr: &str) -> Option<Ipv6Addr> {
+  if let Some(s) = addr.strip_suffix(']') {
+    Ipv6Addr::from_str(s.get(1..)?).ok()
+  } else if let Ok(ip) = IpAddr::from_str(addr) {
+    Some(ip_to_ipv6(ip))
+  } else if let Ok(socket) = SocketAddr::from_str(addr) {
+    Some(ip_to_ipv6(socket.ip()))
+  } else {
+    None
+  }
+}
```

Comment:
Why not use IpAddr FromStr impl from std?

Reply:
`IpAddr::from_str` doesn't handle all possible cases.

* `IpAddr` can only parse an IP address without a port number.

* `SocketAddr` can only parse an IP address combined with a port number.

* Neither `IpAddr` nor `SocketAddr` can parse an IPv6 address with square brackets but no port number.

Reply:
Makes sense. Please put it in a comment, otherwise this info will get lost in the future.